### ISSUE TITLE
List different purposes a test can have

### ISSUE THREAD

**Issue by chillenzer** [edbennett/python-testing-ci#14]
There are quite a few different reasons to test. A definitely incomplete list from the top of my head:

* test for functionality (the first and simplest test that -- in a TDD sense -- is capable of defining a functionality)
* test for confidence/robustness (further tests that are redundant in that they test already tested functionality)
* test for edge cases
* test for failures
* test for ill-formed-input handling
* test for fixing an (external) interface (they trigger if the external interface changes, e.g. due to deprecation etc.)
* test for learning an (external) interface (some tests of an external interface just for learning to work with it)
* ...

Most of them are mentioned in the lesson but it might raise awareness if there is an explicit list somewhere. In particular, there are test purposes that are easily forgotten and not caught by coverage reports or the like. The different purposes of testing have, of course, very different priorities, so you might not always use them but having such a list the learners can consciously decide where they want their code to be located in the continuum between whiteboard-pseudo-code-style quality and high-frequency-billion-dollar-handling software.

**Comment by edbennett** [comment]
Good idea. Where in the lesson would you drop this in? Right at the start?

**Comment by chillenzer** [comment]
I'm not sure. I would love to babble on for an hour in the introduction because I have strong opinions there. But for the sake of the learners, I think it is actually okay that way (was about to suggest a different addition instead but have to make up my mind). As some of them will be covered during the lesson anyways, it might be nice as a concluding remark or a summary? Like the last ten minutes: Present this list, discuss each item and refer to the pertinent part in the previous lesson?
